File: ios/DTO/PreChatSurveyDTO.swift
import CXoneChatSDK
import Foundation

struct PreChatSurveyDTO: Encodable {
  let name: String
  let fields: [FieldDTO]

  init(_ survey: PreChatSurvey) {
    self.name = survey.name
    self.fields = survey.customFields.compactMap { FieldDTO($0) }
  }

  struct FieldDTO: Encodable {
    let id: String
    let label: String
    let required: Bool
    let type: String
    let isEmail: Bool?
    let value: String?
    let options: [OptionDTO]?
    let nodes: [NodeDTO]?

    init?(_ field: PreChatSurveyCustomField) {
      self.required = field.isRequired

      switch field.type {
      case .textField(let textField):
        self.id = textField.ident
        self.label = textField.label
        self.type = textField.isEmail ? "email" : "text"
        self.isEmail = textField.isEmail
        self.value = textField.value
        self.options = nil
        self.nodes = nil

      case .selector(let selector):
        self.id = selector.ident
        self.label = selector.label
        self.type = "select"
        self.isEmail = nil
        self.value = selector.value
        self.options = selector.options.map { OptionDTO(id: $0.key, label: $0.value) }
        self.nodes = nil

      case .hierarchical(let hierarchical):
        self.id = hierarchical.ident
        self.label = hierarchical.label
        self.type = "hierarchical"
        self.isEmail = nil
        self.value = hierarchical.value
        self.options = nil
        self.nodes = hierarchical.nodes.map(NodeDTO.init)
      }
    }
  }

  struct OptionDTO: Encodable {
    let id: String
    let label: String
  }

  struct NodeDTO: Encodable {
    let value: String
    let label: String
    let children: [NodeDTO]

    init(_ node: CustomFieldHierarchicalNode) {
      self.value = node.value
      self.label = node.label
      self.children = node.children.map(NodeDTO.init)
    }
  }
}

File: android/src/main/java/expo/modules/cxonemobilesdk/dto/PreChatSurveyDTO.kt
package expo.modules.cxonemobilesdk.dto

import com.nice.cxonechat.prechat.PreChatSurvey
import com.nice.cxonechat.state.FieldDefinition
import com.nice.cxonechat.state.HierarchyNode
import com.nice.cxonechat.state.SelectorNode
import kotlin.sequences.toList

data class PreChatSurveyDTO(
  val name: String,
  val fields: List<FieldDTO>,
) {
  fun toMap() = mapOf(
    "name" to name,
    "fields" to fields.map { it.toMap() },
  )

  companion object {
    fun from(survey: PreChatSurvey): PreChatSurveyDTO {
      val fieldDtos = survey.fields.toList().mapNotNull { FieldDTO.from(it) }
      return PreChatSurveyDTO(name = survey.name, fields = fieldDtos)
    }
  }

  data class FieldDTO(
    val id: String,
    val label: String,
    val required: Boolean,
    val type: String,
    val value: String?,
    val isEmail: Boolean?,
    val options: List<OptionDTO>?,
    val nodes: List<NodeDTO>?,
  ) {
    fun toMap() = mapOf(
      "id" to id,
      "label" to label,
      "required" to required,
      "type" to type,
      "value" to value,
      "isEmail" to isEmail,
      "options" to options?.map { it.toMap() },
      "nodes" to nodes?.map { it.toMap() },
    )

    companion object {
      fun from(field: FieldDefinition): FieldDTO? = when (field) {
        is FieldDefinition.Text -> FieldDTO(
          id = field.fieldId,
          label = field.label,
          required = field.isRequired,
          type = if (field.isEMail) "email" else "text",
          value = null,
          isEmail = field.isEMail,
          options = null,
          nodes = null,
        )
        is FieldDefinition.Selector -> FieldDTO(
          id = field.fieldId,
          label = field.label,
          required = field.isRequired,
          type = "select",
          value = null,
          isEmail = null,
          options = field.values.toList().map { OptionDTO.from(it) },
          nodes = null,
        )
        is FieldDefinition.Hierarchy -> FieldDTO(
          id = field.fieldId,
          label = field.label,
          required = field.isRequired,
          type = "hierarchical",
          value = null,
          isEmail = null,
          options = null,
          nodes = field.values.toList().map { NodeDTO.from(it) },
        )
        else -> null
      }
    }
  }

  data class OptionDTO(val id: String, val label: String) {
    fun toMap() = mapOf("id" to id, "label" to label)

    companion object {
      fun from(node: SelectorNode): OptionDTO = OptionDTO(
        id = node.nodeId ?: "",
        label = node.label ?: "",
      )
    }
  }

  data class NodeDTO(
    val value: String,
    val label: String,
    val children: List<NodeDTO>,
  ) {
    fun toMap(): Map<String, Any?> {
      val childMaps: List<Map<String, Any?>> = children.map { child -> child.toMap() }
      return mapOf(
        "value" to value,
        "label" to label,
        "children" to childMaps,
      )
    }

    companion object {
      fun from(node: HierarchyNode<String>): NodeDTO {
        val childNodes: List<NodeDTO> = node.children
          .map { child -> from(child) }
          .toList()
        return NodeDTO(
          value = node.nodeId ?: "",
          label = node.label ?: "",
          children = childNodes,
        )
      }
    }
  }
}
